File: kwj1270/4월 5주차_찐/보석쇼핑.cpp
#include <string>
#include <vector>
#include <set>
#include <map>

using namespace std;

vector<int> solution(vector<string> gems) {
    vector<int> answer = { 1, (int)gems.size() };
    map<string, int> m;
    set<string> s(gems.begin(), gems.end());
    int start = 0;
    int end = 0;
    int ans = gems.size();
    while(true){
        if(m.size() == s.size()){
            if(ans > end-start){
                ans = end-start;
                answer[0] = start+1;
                answer[1] = end;
            }
            if(m[gems[start]] == 1){
                m.erase(gems[start]);
                start++;
            }
            else if(m[gems[start]]-1 > 0){
                m[gems[start]]--;
                start++;
            }
        }else if(end == gems.size() || start == gems.size()){
            break;
        }else{
            m[gems[end]]++; // 개수 하나 증가시키고
            end++; // 다음 위치로 이동
        }
    }
    return answer;
}

File: akrnrh3737/4월 5주차/보석쇼핑.py
from collections import defaultdict

def solution(gems):
    answer = [] # 길이, [시작인덱스, 끝인덱스]
    countGems = defaultdict(int)
    countGems[gems[0]]=1
    typeCnt = len(set(gems)) # 보석 종류 개수
    length = len(gems) # 진열대 길이

    start,end = 0,0
    while start < length and end < length:
        if len(countGems) < typeCnt: # 아직 보석 다 못삼
            end+=1 #오른쪽으로 이동
            if end == length: 
                break
            countGems[gems[end]]+=1 # 이동했으니깐 보석 빈도수 증가

        # 조건 만족함
        if len(countGems) == typeCnt:
            answer.append((end-start, [start+1, end+1]))

            #왼쪽빼주기
            countGems[gems[start]]-=1
            if countGems[gems[start]] == 0:
                del countGems[gems[start]]
            start += 1

    answer = sorted(answer, key=lambda x : (x[0], x[1])) # 길이가 같으면 인덱스가 작게
    return answer[0][1]

if __name__ == "__main__":
    gems = ["DIA", "RUBY", "RUBY", "DIA", "DIA", "EMERALD", "SAPPHIRE", "DIA"]
    print(solution(gems))